xgcm/grid.py

```diff
@@ -1162,29 +1183,49 @@ def integrate(self, da, axis, **kwargs):
         da_i : xarray.DataArray
             The integrated data
         """
-        check_axes = [ax for ax in axis if ax not in self.axes]
-        check_axes_str = ",".join(check_axes)
-        if any(check_axes):
-            raise ValueError("Axis %s not found in grid object" % check_axes_str)
-
-        dim = []
-        for ax in axis:
-            all_dim = self.axes[ax].coords.values()
-            matching_dim = [di for di in all_dim if di in da.dims]
-            if len(matching_dim) == 1:
-                dim.append(matching_dim[0])
-            else:
-                raise ValueError(
-                    "Did not find single matching dimension corresponding to axis %s. Got (%s)"
-                    % (ax, matching_dim)
-                )
+        # check if axis input is compatible
+        self._check_axes(axis)
 
+        # get dimension(s) corresponding
+        # to `da` and `axis` input
+        dim = self._get_dims_from_axis(da, axis)
+        print(dim)
         weight = self.get_metric(da, axis)
         weighted = da * weight
         # TODO: We should integrate xarray.weighted once available.
 
         return weighted.sum(dim)
 
+    @docstrings.dedent
+    def average(self, da, axis, **kwargs):
+        """
+        Perform weighted average along specified axis or axes,
+        accounting for grid metrics. (e.g. cell length, area, volume)
+
+        Parameters
+        ----------
+        axis : str, list of str
+            Name of the axis on which to act
+        %(neighbor_binary_func.parameters.no_f)s
+
+        Returns
+        -------
+        da_i : xarray.DataArray
+            The averaged data
+        """
+        # check if axis input is compatible
+        self._check_axes(axis)
```

Comment:
Doesn't this check happen in `get_metric` anyway? Why do we need to call it twice?

Reply:
Oh you might be right. Ill investigate

Reply:
Ok i have dug a little. Consider this small example:
```
from xgcm.test.datasets import datasets_grid_metric
from xgcm import Grid

axis = 'X'
funcname = 'integrate'
ds, coords, metrics = datasets_grid_metric("C")

del coords[axis]

del_metrics = [k for k in metrics.keys() if axis in k]
for dm in del_metrics:
    del metrics[dm]

grid = Grid(ds, coords=coords, metrics=metrics)

func = getattr(grid, funcname)

grid.get_metric(ds.u, ['X'])
```

This does indeed raise an error, but it is somewhat cryptic: 
```
"Unable to find any combinations of metrics for array dims {'yt', 'time', 'xu', 'zt'} and axes ['X']"
```

So while I agree that the checking should probably be removed, should we work on the error here?

Reply:
We can discuss this, for now I will remove the additional function and check that this error is properly raised in the tests.